modules/navier_stokes/src/materials/INSFVMushyPorousFrictionMaterial.C

```diff
@@ -0,0 +1,66 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "INSFVMushyPorousFrictionMaterial.h"
+#include "MooseMesh.h"
+#include "NS.h"
+
+registerMooseObject("NavierStokesApp", INSFVMushyPorousFrictionMaterial);
+
+InputParameters
+INSFVMushyPorousFrictionMaterial::validParams()
+{
+  InputParameters params = FunctorMaterial::validParams();
+  params.addClassDescription(
+      "Computes the mushy zone porous resistance for solidification/melting problems.");
+  params.addRequiredParam<MooseFunctorName>("liquid_fraction", "Liquid Fraction Functor.");
+  params.addRequiredParam<MooseFunctorName>(NS::mu, "The liquid dynamic viscosity.");
+  params.addRequiredParam<MooseFunctorName>("rho_l", "The liquid density (not the mixture one).");
+  params.addParam<MooseFunctorName>(
+      "dendrite_spacing_scaling", "1e-4", "The dendrite spacing scaling.");
+  return params;
+}
+
+INSFVMushyPorousFrictionMaterial::INSFVMushyPorousFrictionMaterial(
+    const InputParameters & parameters)
+  : FunctorMaterial(parameters),
+    _fl(getFunctor<ADReal>("liquid_fraction")),
+    _mu(getFunctor<ADReal>(NS::mu)),
+    _rho_l(getFunctor<ADReal>("rho_l")),
+    _dendrite_spacing_scaling(getFunctor<ADReal>("dendrite_spacing_scaling"))
+{
+
+  addFunctorProperty<ADReal>(
+      "Darcy_coefficient",
+      [this](const auto & r, const auto & t) -> ADReal
+      {
+        constexpr Real epsilon = 1e-15; // prevents explosion of sqrt(x) derivative to infinity
+
+        const auto fs = 1.0 - _fl(r, t);
+        const auto cs = _c / Utility::pow<2>(_dendrite_spacing_scaling(r, t));
+        const auto Fk = 0.5 + std::atan(_s * (fs - _fs_crit)) / pi;
+        const auto K =
+            Utility::pow<3>(_fl(r, t)) / ((Utility::pow<2>(fs) + epsilon) * Fk * cs) + epsilon;
```

Comment:
same before dividing

Reply:
I thought we had the epsilon there for this reason.

Reply:
doesnt cover Fk=0 (hard to be there) and cs = 0

Reply:
Considering that `_c` and `_s` are positive numbers (not input parameters) and `Fk` has a `+0.5` before the arctan whose argument is also bounded between 0 and 1, I think no need to assert here. Considering that we assert that dendrite spacing is not 0.